protocolupgrade/engine.go

```diff
@@ -0,0 +1,243 @@
+// of this software will be governed by version 3 or later of the GNU General
+// Public License.
+
+package protocolupgrade
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sort"
+	"sync"
+
+	vegapb "code.vegaprotocol.io/protos/vega"
+	commandspb "code.vegaprotocol.io/protos/vega/commands/v1"
+	eventspb "code.vegaprotocol.io/protos/vega/events/v1"
+	"code.vegaprotocol.io/vega/events"
+	"code.vegaprotocol.io/vega/logging"
+	"code.vegaprotocol.io/vega/txn"
+	"code.vegaprotocol.io/vega/types"
+	"code.vegaprotocol.io/vega/types/num"
+	"github.com/blang/semver"
+	"github.com/cenkalti/backoff"
+	"github.com/golang/protobuf/proto"
+)
+
+var requiredQuorum = num.NewDecimalFromFloat(2).Div(num.NewDecimalFromFloat(3))
+
+type protocolUpgradeProposal struct {
+	blockHeight uint64
+	tag         string
+	accepted    map[string]struct{}
+}
+
+func protocolUpgradeProposalID(upgradeBlockHeight uint64, releaseTag string) string {
+	return fmt.Sprintf("%v@%v", releaseTag, upgradeBlockHeight)
+}
+
+func (p *protocolUpgradeProposal) approvers() []string {
+	accepted := make([]string, 0, len(p.accepted))
+	for k := range p.accepted {
+		accepted = append(accepted, k)
+	}
+	sort.Strings(accepted)
+	return accepted
+}
+
+type ValidatorTopology interface {
+	IsTendermintValidator(pubkey string) bool
+	IsSelfTendermintValidator() bool
+	NumberOfTendermintValidators() uint
+}
+
+type Commander interface {
+	CommandSync(ctx context.Context, cmd txn.Command, payload proto.Message, f func(error), bo *backoff.ExponentialBackOff)
+}
+
+type Broker interface {
+	Send(event events.Event)
+}
+
+type Engine struct {
+	log      *logging.Logger
+	config   Config
+	broker   Broker
+	topology ValidatorTopology
+	cmd      Commander
+	hashKeys []string
+
+	currentBlockHeight    uint64
+	activeProposals       map[string]*protocolUpgradeProposal
+	events                map[string]*eventspb.ProtocolUpgradeEvent
+	blockHeightForUpgrade int
+	readyForUpgrade       bool
+	lock                  sync.RWMutex
+	ctx                   context.Context
+}
+
+func New(log *logging.Logger, config Config, broker Broker, topology ValidatorTopology, cmd Commander) *Engine {
+	log = log.Named(namedLogger)
+	log.SetLevel(config.Level.Get())
+	return &Engine{
+		activeProposals:       map[string]*protocolUpgradeProposal{},
+		events:                map[string]*eventspb.ProtocolUpgradeEvent{},
+		blockHeightForUpgrade: -1,
+		log:                   log,
+		config:                config,
+		broker:                broker,
+		topology:              topology,
+		cmd:                   cmd,
+		hashKeys:              []string{(&types.PayloadProtocolUpgradeProposals{}).Key()},
+	}
+}
+
+// UpgradeProposal records the intention of a validator to upgrade the protocol to a release tag at block height.
+func (e *Engine) UpgradeProposal(ctx context.Context, pk string, upgradeBlockHeight uint64, releaseTag string) error {
+	if !e.topology.IsTendermintValidator(pk) {
```

Comment:
Not sure what that actually means but would that not be an issue for non-validators (maybe those who runs data-node)?